src/hypertable_restrict_info.c

```diff
@@ -71,68 +73,128 @@ dimension_restrict_info_create(Dimension *d)
-			}
-			return true;
-		case BTGreaterEqualStrategyNumber:
-		case BTGreaterStrategyNumber:
-			if (dri->lower_strategy == InvalidStrategy || value > dri->lower_bound)
-			{
-				dri->lower_strategy = strategy;
+		Datum		datum = PointerGetDatum(lfirst(item));
+		int64		value = time_value_to_internal(datum, type, true);
+
+		if (value == -1)
+		{
+			/* open dimension does not support IN/ANY/ALL restrictions */
+			elog(WARNING, "unknown time restriction type %d", type);
+			return restriction_added;
+		}
+
+		switch (strategy)
+		{
+			case BTLessEqualStrategyNumber:
+			case BTLessStrategyNumber:
+				if (dri->upper_strategy == InvalidStrategy || value < dri->upper_bound)
+				{
+					dri->upper_strategy = strategy;
+					dri->upper_bound = value;
+					restriction_added = true;
+				}
+				break;
+			case BTGreaterEqualStrategyNumber:
+			case BTGreaterStrategyNumber:
+				if (dri->lower_strategy == InvalidStrategy || value > dri->lower_bound)
+				{
+					dri->lower_strategy = strategy;
+					dri->lower_bound = value;
+					restriction_added = true;
+				}
+				break;
+			case BTEqualStrategyNumber:
 				dri->lower_bound = value;
-			}
-			return true;
-		case BTEqualStrategyNumber:
-			dri->lower_bound = value;
-			dri->upper_bound = value;
-			dri->lower_strategy = BTGreaterEqualStrategyNumber;
-			dri->upper_strategy = BTLessEqualStrategyNumber;
-			return true;
-		default:
-			return false;
+				dri->upper_bound = value;
+				dri->lower_strategy = BTGreaterEqualStrategyNumber;
+				dri->upper_strategy = BTLessEqualStrategyNumber;
+				restriction_added = true;
+				break;
+			default:
+				/* unsupported strategy */
+				break;
+
+		}
 	}
+	return restriction_added;
 }
 
 static bool
-dimension_restrict_info_closed_add(DimensionRestrictInfoClosed *dri, StrategyNumber strategy, Const *c)
+dimension_restrict_info_closed_add(DimensionRestrictInfoClosed *dri, StrategyNumber strategy, List *values, bool use_or)
 {
-	int64		value = partitioning_func_apply(dri->base.dimension->partitioning, c->constvalue);
+	ListCell   *item;
+	bool		restriction_added = false;
 
-	switch (strategy)
+	if (strategy != BTEqualStrategyNumber)
 	{
-		case BTEqualStrategyNumber:
-			dri->value = value;
+		return restriction_added;
+	}
+
+	if (dri->partitions == NIL)
+	{
+		foreach(item, values)
+		{
+			Datum		datum = PointerGetDatum(lfirst(item));
+			int32		partition = partitioning_func_apply(dri->base.dimension->partitioning, datum);
+
+			dri->partitions = list_append_unique_int(dri->partitions, partition);
 			dri->strategy = strategy;
-			return true;
-		default:
-			return false;
+			restriction_added = true;
+
+			if (remove_all_partitions(use_or, dri))
+				return restriction_added;
+		}
 	}
+	else
+	{
+		List	   *tmp = NIL;
+
+		foreach(item, values)
+		{
+			Datum		datum = PointerGetDatum(lfirst(item));
+			int32		partition = partitioning_func_apply(dri->base.dimension->partitioning, datum);
+
+			tmp = lappend_int(tmp, partition);
```

Comment:
Yes, since we iterate over unique `dri->partitions` and check existence in `tmp`. I've skipped unique check here because current PostgreSQL impl. is quadratic time complexity